### PR TITLE
Issue 258 basexppauli class skeleton

### PR DESCRIPTION
**Pull request by dhruvbhq**
### Summary
Fix for issue #258 . Added skeleton class `BaseXPPauli` and methods `xp_pauli_rep.py`

### Details and comments
- Added skeleton class and methods in qiskit_qec/operators and qiskit_qec/utils
- Most function declarations from `BasePauli` and `pauli_rep.py` have been retained as of now. Functions are empty. There are placeholders for future documentation.
- Locally disabled lint warnings for newly added files
- Fixed a typo

### CODE REVIEW

File: qiskit_qec/operators/base_xp_pauli.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+from qiskit.circuit.instruction import Instruction
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit.quantum_info.operators.mixins import AdjointMixin, MultiplyMixin
+from qiskit_qec.utils import xp_pauli_rep
+
+
+# pylint: disable=no-member
+class BaseXPPauli(BaseOperator, AdjointMixin, MultiplyMixin):
+    r"""Symplectic representation of a list of N-qubit XP operators with phases using
+    numpy arrays for symplectic matrices and phase vectors.
```
Comment (dsvandet) [inline qiskit_qec/operators/base_xp_pauli.py:36]: ... using numpy arrays for generalized symplectic matrices ...

File: qiskit_qec/operators/base_xp_pauli.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+    EXTERNAL_QUBIT_ORDER = xp_pauli_rep.DEFAULT_QUBIT_ORDER
+
+    PRINT_PHASE_ENCODING = None
+
+    def __init__(
+        self,
+        matrix: Union[np.ndarray, None] = None,
+        phase_exp: Union[None, np.ndarray, np.integer] = None,
+        order: str = "xz",
+    ) -> None:
```
Comment (dsvandet) [inline qiskit_qec/operators/base_xp_pauli.py:56]: A precision variable N is needed for XP Operators

File: qiskit_qec/operators/base_xp_pauli.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+        """A BaseXPPauli object represents a list N-qubit Pauli operators with phases.
```
Comment (dsvandet) [inline qiskit_qec/operators/base_xp_pauli.py:57]: A BaseXPPauli object represents a list N-qubit **XP** Pauli operators

File: qiskit_qec/operators/base_xp_pauli.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+        Numpy arrays are used to represent the symplectic matrix represention of these
+        Paulis. The phases of the Paulis are stored encoded. The phases of the Pauli
```
Comment (dsvandet) [inline qiskit_qec/operators/base_xp_pauli.py:59]: Leave a TODO here as we need to work out the representations to be used for the XP operators. Also the

` (-i)^phase_exp Z^z X^x` 
 
 needs to be updated to XP operators etc.

**Comment by dhruvbhq** [comment]
@dsvandet Thanks, I implemented the suggestions. Just one concern - `N` is used for the precision variable, at the same time, many docstrings say something like `...N-qubit Pauli operator...` etc. Could it be a potential source of confusion?

**Comment by dsvandet** [comment]
Good point. Let's call it `precision` instead of `N`.

**Comment by dsvandet** [comment]
In terms of the naming PauliXP  there are cases where  the XP operators no longer include the original Paulis. I view the XP operators  as 'generalized Paulis' in terms of not what the specifically are but in terms of their role and definition. From this point of view we do not need to always being including the original Paulis. In most cases the precision is going to be chosen such that the original Pauls are included. So let's leave this for the moment. We can always change later if a good reason shows itself.

**Comment by dhruvbhq** [comment]
Thanks for your answer @dsvandet , it makes a lot of sense to me.
I've renamed `N` to `precision` and also changed the datatype `np.int8` to `np.int64` in `BaseXPPauli`, as discussed.

**Review by grace-harper-ibm — APPROVED** [review]
LGTM

**Review by dsvandet — APPROVED** [review]
We will want to revisit the use of np.integer. The reason is that np.integer can be very slow to check. Not important at this point so let's not change it now.

**Pull request merged**